### PR TITLE
[flang][OpenMP] Parsing support for map type modifiers

### PR DESCRIPTION
**Pull request by kparzysz**
This commit adds parsing of type modifiers for the MAP clause: CLOSE, OMPX_HOLD, and PRESENT. The support for ALWAYS has already existed.

The new modifiers are not yet handled in lowering: when present, a TODO message is emitted and compilation stops.

### CODE REVIEW

File: flang/lib/Parser/unparse.cpp [hunk 1]
```diff
@@ -2073,11 +2073,24 @@ class UnparseVisitor {
...
+    if (typeMod.has_value() || type.has_value())
+      Put(": ");
     Walk(std::get<OmpObjectList>(x.t));
   }
+  void Unparse(const OmpMapClause::TypeModifier &x) {
+    if (x == OmpMapClause::TypeModifier::OmpxHold)
+      Word("OMPX_HOLD");
+    else
+      Word(OmpMapClause::EnumToString(x));
+  }
```
Comment (kiranchandramohan) [inline flang/lib/Parser/unparse.cpp:2093]: Nit: Frontend code has braces almost always.

File: flang/lib/Parser/unparse.cpp [hunk 2]
```diff
@@ -2073,11 +2073,24 @@ class UnparseVisitor {
...
+    if (typeMod.has_value() && type.has_value())
+      Put(", ");
+    Walk(type);
+    if (typeMod.has_value() || type.has_value())
+      Put(": ");
     Walk(std::get<OmpObjectList>(x.t));
   }
+  void Unparse(const OmpMapClause::TypeModifier &x) {
+    if (x == OmpMapClause::TypeModifier::OmpxHold)
+      Word("OMPX_HOLD");
```
Comment (kiranchandramohan) [inline flang/lib/Parser/unparse.cpp:2090]: Why is this a special case?

File: flang/lib/Lower/OpenMP/ClauseProcessor.cpp [hunk 3]
```diff
@@ -945,40 +945,42 @@ bool ClauseProcessor::processMap(
...
+        auto &modTypeMods =
+            std::get<std::optional<Map::MapTypeModifiers>>(clause.t);
+        if (modTypeMods) {
+          if (llvm::is_contained(*modTypeMods, Map::MapTypeModifier::Always))
+            mapTypeBits |= llvm::omp::OpenMPOffloadMappingFlags::OMP_MAP_ALWAYS;
+          // Diagnose unimplemented map-type-modifiers.
+          if (llvm::any_of(*modTypeMods, [](Map::MapTypeModifier m) {
+                return m != Map::MapTypeModifier::Always;
+              })) {
+            TODO(currentLocation, "Map type modifiers (other than 'ALWAYS')"
```
Comment (kiranchandramohan) [inline flang/lib/Lower/OpenMP/ClauseProcessor.cpp:981]: Can you add a test for the TODO?

**Review by kiranchandramohan — APPROVED** [review]
LGTM. Thanks Kryztof.

File: flang/lib/Parser/unparse.cpp (hunk 1 above)
Comment (kparzysz) [reply flang/lib/Parser/unparse.cpp]: Done

File: flang/lib/Lower/OpenMP/ClauseProcessor.cpp (hunk 3 above)
Comment (kparzysz) [reply flang/lib/Lower/OpenMP/ClauseProcessor.cpp:981]: Done

**Pull request merged**